### DISCUSSION TITLE
Custom implementation of Imageprovider with cache

### DISCUSSION THREAD

**Discussion by jbppraxis** [SixLabors/ImageSharp.Web#338] [Q&A]
We have created a new imageservice with a custom imageprovider which fetches images from a database.
Furthermore we have a imageprovider fetching images from a blobstorage.
Also we have added a azure blobstorage cache as well.

Our startup
```
builder.Services.AddImageSharp()
        .ClearProviders()
        .SetCache<AzureBlobStorageCache>()
        .Configure<AzureBlobStorageCacheOptions>(options =>
        {
            options.ConnectionString = builder.Configuration["xx"];
            options.ContainerName = "cache";

            // Optionally create the cache container on startup if not already created.
            AzureBlobStorageCache.CreateIfNotExists(options, PublicAccessType.None);
        })
        .Configure<AzureBlobStorageImageProviderOptions>(options =>
        {
            // The "BlobContainers" collection allows registration of multiple containers.
            options.BlobContainers.Add(new AzureBlobContainerClientOptions
            {
                ConnectionString = builder.Configuration["xx"],
                ContainerName = "xx",
            });
        })
        .AddProvider<AzureBlobStorageImageProvider>()
        .AddProvider<DatabaseImageProvider>() ;
```
We can fetch images from both the database and the blobstorage.
We can see that the image is placed in the blob storage cache with a hashed filename.

However we are not sure that the image from the cache is the one being served.

When adding a breakpoint to the GetAsync() implementation of our custom imageprovider we can see that is is invoked on every request.
I would expect the GetAsync method of the imageprovider only to be invoked if the image is  missing from the cache or when the cacheControlMaxAge is reached.

In the imageresolver The metadata of the is set in the following way  `_imageMetaData = new ImageMetadata(image.UploadedDate.ToUniversalTime(), TimeSpan.FromDays(365), image.RawImage.Length);`

Any suggestion to what we have done wrong or is missing is highly appreciated.

**Comment by JimBobSquarePants** [comment]
I can only really suggest that you pull down the source here and add your custom provider to the sample project. That will allow you to debug the entire pipeline.
